fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java

```diff
@@ -235,6 +235,13 @@ private void buildUniqueList()
         LOGGER.debug(LOADING, "Found {} mod requirements missing ({} mandatory, {} optional)", missingVersions.size(), mandatoryMissing, missingVersions.size() - mandatoryMissing);
 
         if (!missingVersions.isEmpty()) {
+            if (mandatoryMissing > 0) {
+                LOGGER.error(LOADING, "Missing mandatory dependencies: {}", missingVersions.stream().filter(IModInfo.ModVersion::isMandatory).map(IModInfo.ModVersion::getModId).collect(Collectors.joining(", ")));
+            }
+            if (missingVersions.size() - mandatoryMissing > 0) {
+                LOGGER.error(LOADING, "Missing optional dependencies: {}", missingVersions.stream().filter(ver -> !ver.isMandatory()).map(IModInfo.ModVersion::getModId).collect(Collectors.joining(", ")));
```

Comment:
Upon further thought, this should be better worded as `"Unsupported installed optional dependencies"`, as "missing" implies optional dependencies must exist, but in this case we mean "optional dependency which is present but falls outside of the version range.

Reply:
That is definitely better, fixed